common/djangoapps/student/models.py

```diff
@@ -1894,29 +1894,64 @@ def refundable(self, user_already_has_certs_for=None):
+            username = self.user.username
+            enrollment_attributes = get_enrollment_attributes(username, str(self.course_id))
+            enrollment_attributes.append(
+                {
+                    "namespace": "order",
+                    "name": "date_placed",
+                    "value": date_placed,
+                }
+            )
+            set_enrollment_attributes(username, str(self.course_id), enrollment_attributes)
+
+        refund_window_start_date = max(
+            datetime.strptime(date_placed, ECOMMERCE_DATE_FORMAT),
+            self.course_overview.start.replace(tzinfo=None)
+        )
+
+        return refund_window_start_date.replace(tzinfo=UTC) + EnrollmentRefundConfiguration.current().refund_window
+
+    def is_order_voucher_refundable(self):
+        """ Checks if the coupon batch expiration date has passed to determine whether order voucher is refundable. """
+        from openedx.core.djangoapps.commerce.utils import ECOMMERCE_DATE_FORMAT
+        vouchers = self.get_order_attribute_from_ecommerce('vouchers')
+        if not vouchers:
+            return False
+        voucher_end_datetime_str = vouchers[0]['end_datetime']
+        voucher_expiration_date = datetime.strptime(voucher_end_datetime_str, ECOMMERCE_DATE_FORMAT).replace(tzinfo=UTC)
+        return datetime.now(UTC) < voucher_expiration_date
+
+    def get_order_attribute_from_ecommerce(self, attribute_name):
+        """Fetches the order details from ecommerce to return the value of the attribute passed as argument."""
+        # NOTE: This is here to avoid circular references
+        from openedx.core.djangoapps.commerce.utils import ecommerce_api_client
+        order_number = self.get_order_attribute_value('order_number')
+        if not order_number:
+            return None
+
+        # check if response is already cached
+        cache_key = get_cache_key(user_id=self.user.id, order_number=order_number)
+        cached_response = TieredCache.get_cached_response(cache_key)
+        if cached_response.is_found:
+            order = cached_response.value
+        else:
             try:
+                # response is not cached, so make a call to ecommerce to fetch order details
                 order = ecommerce_api_client(self.user).orders(order_number).get()
-                date_placed = order['date_placed']
-                # also save the attribute so that we don't need to call ecommerce again.
-                username = self.user.username
-                enrollment_attributes = get_enrollment_attributes(username, str(self.course_id))
-                enrollment_attributes.append(
-                    {
-                        "namespace": "order",
-                        "name": "date_placed",
-                        "value": date_placed,
-                    }
-                )
-                set_enrollment_attributes(username, str(self.course_id), enrollment_attributes)
```

Comment:
Why we have moved the above deleted code into `refund_cutoff_date`?

Reply:
All this code was actually already a part of `refund_cutoff_date`, I haven't deleted or moved it into `refund_cutoff_date`, I have only divided the logic into two methods now. A new generic method `get_order_attribute_from_ecommerce` is added which takes care of fetching order details from e-commerce 
and the rest of the logic stays in `refund_cutoff_date` as it there was. 

The reason why I have created this generic method is to avoid repetition, otherwise I would have to write this same code to fetch voucher details in `is_order_voucher_refundable`. Hope that answers your concerns, if not, I can explain it on call.

Reply:
Adding original snippet of `refund_cutoff_date` here to clear the confusion:

```python
    def refund_cutoff_date(self):
        """ Calculate and return the refund window end date. """
        # NOTE: This is here to avoid circular references
        from openedx.core.djangoapps.commerce.utils import ecommerce_api_client, ECOMMERCE_DATE_FORMAT

        date_placed = self.get_order_attribute_value('date_placed')

        if not date_placed:
            order_number = self.get_order_attribute_value('order_number')
            if not order_number:
                return None

            try:
                order = ecommerce_api_client(self.user).orders(order_number).get()
                date_placed = order['date_placed']
                # also save the attribute so that we don't need to call ecommerce again.
                username = self.user.username
                enrollment_attributes = get_enrollment_attributes(username, str(self.course_id))
                enrollment_attributes.append(
                    {
                        "namespace": "order",
                        "name": "date_placed",
                        "value": date_placed,
                    }
                )
                set_enrollment_attributes(username, str(self.course_id), enrollment_attributes)
            except HttpClientError:
                log.warning(
                    "Encountered HttpClientError while getting order details from ecommerce. "
                    "Order={number} and user {user}".format(number=order_number, user=self.user.id))
                return None

            except HttpServerError:
                log.warning(
                    "Encountered HttpServerError while getting order details from ecommerce. "
                    "Order={number} and user {user}".format(number=order_number, user=self.user.id))
                return None

            except SlumberBaseException:
                log.warning(
                    "Encountered an error while getting order details from ecommerce. "
                    "Order={number} and user {user}".format(number=order_number, user=self.user.id))
                return None

        refund_window_start_date = max(
            datetime.strptime(date_placed, ECOMMERCE_DATE_FORMAT),
            self.course_overview.start.replace(tzinfo=None)
        )

        return refund_window_start_date.replace(tzinfo=UTC) + EnrollmentRefundConfiguration.current().refund_window
```